#644 add visitor property to both PP and PPC

xiewenya opened the pull request.

xiewenya committed:
#644 add visitor property to both PP and PPC

xiewenya committed:
#644 add visitor property to both PP and PPC

xiewenya committed:
#644 add visitor property to both PP and PPC

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrinter.java:
```diff
@@ -28,18 +28,24 @@
  */
 public class PrettyPrinter {
     private final PrettyPrinterConfiguration configuration;
+    private PrettyPrintVisitor visitor;
```
If the visitor is in the configuration, then we don't really need to add it here

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrinter.java:
```diff
@@ -28,18 +28,24 @@
     }
 
     public String print(Node node) {
         final PrettyPrintVisitor visitor = new PrettyPrintVisitor(configuration);
```
We can move this initialization of visitor the the field in PPConfiguration. Then we just call get on the configuration here.

matozoid replied:
Oh I see, the visitor needs the configuration :(

matozoid replied:
Damn this mutability!

matozoid replied:
So I guess this can be solved by setting the configuration on the visitor when it is initialized or set in the configuration. So that will need a setter on the visitor.

xiewenya committed:
add VisitorClass property in PPC
create new instance of visitor using relfection in PP
make SourcePrinter in PP protected

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrinterConfiguration.java:
```diff
@@ -24,6 +24,7 @@
 public class PrettyPrinterConfiguration {
     private boolean printComments = true;
     private String indent = "    ";
+    private Class<? extends PrettyPrintVisitor> visitorClass;
```
set this to new PrettyPrinterVisitor() to get a default

matozoid replied:
No wait, you made it a class. Don't make it a class. We don't need a new instance for every print.

xiewenya replied:
the previous version also create new instance for each call, did I miss something?

```
 public String print(Node node) {
        final PrettyPrintVisitor visitor = new PrettyPrintVisitor(configuration);
        node.accept(visitor, null);
        return visitor.getSource();
    }
```

And if the visitor is shared among different print call, test cases fails.
For example, if I add a visitor to PP class, and pass it to node.accept()
```
public class PrettyPrinter {
    private final PrettyPrinterConfiguration configuration;
    private PrettyPrintVisitor visitor;

    public PrettyPrinter() {
        this(new PrettyPrinterConfiguration());
    }

    public PrettyPrinter(PrettyPrinterConfiguration configuration) {
        this.configuration = configuration;
        visitor = new PrettyPrintVisitor(configuration);
    }

    public String print(Node node) {
        node.accept(visitor, null);
        return visitor.getSource();
    }
}
```

For test case `retainAll()`, it reports

    java.lang.AssertionError: 
    Expected :['int a;' REMOVAL in list at 0, 'int c;' REMOVAL in list at 1, 'int e;' REMOVAL in list at 2]
    Actual   :['intint aint;' REMOVAL in list at 0, 'intint aint;intint cint;' REMOVAL in list at 1, 'intint aint;intint cint;intint eint;' REMOVAL in list at 2]

That‘s why I used reflection. So I'm a little confused here, maybe we can discuss it on gitter?

matozoid replied:
Hmmm, I see. Sure, let's go to gitter.

matozoid replied:
After some thinking: maybe an overridable factory method on the configuration would be best, like "createVisitor(config)"

matozoid committed:
Merge branch 'master' into master

xiewenya committed:
add createVisit to PPC

xiewenya committed:
Merge remote-tracking branch 'origin/master'

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrintVisitor.java:
```diff
@@ -51,8 +51,7 @@
  */
 public class PrettyPrintVisitor implements VoidVisitor<Void> {
     private final PrettyPrinterConfiguration configuration;
```
I guess you can make this protected too

matozoid commented on javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrinterConfiguration.java:
```diff
@@ -42,4 +45,20 @@ public PrettyPrinterConfiguration setPrintComments(boolean printComments) {
+
+    public PrettyPrintVisitor createVisitor() {
+        try {
+            Constructor<? extends PrettyPrintVisitor> constructor =
```
Please no reflection like this - there is nothing here that requires us to subvert type safety. Why not configure things with a Function<PPC, PPV> ?

xiewenya replied:
Did you mean using generics as follows or moving createVisitor to an extra factory class? Could you be more specific?
```
public class PrettyPrinterConfiguration<T extends PrettyPrintVisitor> {
    private Class<T> visitorClass = (Class<T>) PrettyPrintVisitor.class;

    public T createVisitor() {
        try {
            Constructor<T> constructor =
                    visitorClass.getConstructor(PrettyPrinterConfiguration.class);
            return constructor.newInstance(this);
        } catch (Exception e) {
            assert false;
            return null;
        }
    }
}
```

matozoid committed:
Use a factory function

xiewenya committed:
Merge pull request #1 from matozoid/xiewenya_pretty_printer

Use a factory function

xiewenya committed:
Merge branch 'master' into master

xiewenya committed:
made configuration in PPV protected
made class SourcePrinter public

matozoid committed:
Merge branch 'master' into master
